AwsCryptographicMaterialProviders/dafny/AwsCryptographicMaterialProviders/src/CMMs/CachingCMM.dfy

```diff
@@ -0,0 +1,1100 @@
+              == Primitives.Types.DigestInput(
+                digestAlgorithm := Primitives.Types.SHA_512,
+                message :=
+                  partitionKeyDigest
+                  + [0x01]
+                  + AlgorithmSuites.GetSuite(input.algorithmSuiteId.value).binaryId
+                  + ecDigestRequest.output.value
+              )
+            else
+              Seq.Last(cryptoPrimitives.History.Digest).input
+              == Primitives.Types.DigestInput(
+                digestAlgorithm := Primitives.Types.SHA_512,
+                message :=
+                  partitionKeyDigest
+                  + [0x00]
+                  + ecDigestRequest.output.value
+              )
+           )
+        && Seq.Last(cryptoPrimitives.History.Digest).output.value == output.value
+  {
+    var ecDigest :- Sha512EncryptionContext(input.encryptionContext, cryptoPrimitives);
+
+    var output';
+
+    if input.algorithmSuiteId.Some? {
+      var algorithmSuiteBinaryId := AlgorithmSuites.GetSuite(input.algorithmSuiteId.value).binaryId;
+      output' := cryptoPrimitives.Digest(
+        Primitives.Types.DigestInput(
+          digestAlgorithm := Primitives.Types.SHA_512,
+          message :=
+            partitionKeyDigest
+            + [0x01]
+            + algorithmSuiteBinaryId
+            + ecDigest
+        ));
+    } else {
+      output' := cryptoPrimitives.Digest(
+        Primitives.Types.DigestInput(
+          digestAlgorithm := Primitives.Types.SHA_512,
+          message :=
+            partitionKeyDigest
+            + [0x00]
+            + ecDigest
+        ));
+    }
+    output := output'.MapFailure(e => Types.AwsCryptographyPrimitives(e));
+  }
+
+  const PADDING_OF_512_ZERO_BITS: seq<uint8> := [
+    0,0,0,0,0,0,0,0,
+    0,0,0,0,0,0,0,0,
+    0,0,0,0,0,0,0,0,
+    0,0,0,0,0,0,0,0,
+    0,0,0,0,0,0,0,0,
+    0,0,0,0,0,0,0,0,
+    0,0,0,0,0,0,0,0,
+    0,0,0,0,0,0,0,0
+  ]
+
+  lemma PaddingIsCorrect()
+    ensures |PADDING_OF_512_ZERO_BITS| == 64
+    ensures 64 * 8 == 512
+    ensures forall b <- PADDING_OF_512_ZERO_BITS :: b == 0
+  {}
+
+  method {:vcs_split_on_every_assert} DecryptMaterialsCacheKey(
+    partitionKeyDigest: seq<uint8>,
+    input: Types.DecryptMaterialsInput,
+    cryptoPrimitives: Primitives.AtomicPrimitivesClient
+  )
+    returns (output: Result<seq<uint8>, Types.Error>)
+    requires cryptoPrimitives.ValidState()
+    modifies cryptoPrimitives.Modifies
+    ensures cryptoPrimitives.ValidState()
+
+    ensures
+      output.Success?
+      ==>
+        && |cryptoPrimitives.History.Digest| == |old(cryptoPrimitives.History.Digest)| + |input.encryptedDataKeys| + 2
+        && var outputRequest := Seq.Last(cryptoPrimitives.History.Digest);
+        && var edkDigestRequests
+          := cryptoPrimitives.History.Digest[
+             |old(cryptoPrimitives.History.Digest)| .. |old(cryptoPrimitives.History.Digest)| + |input.encryptedDataKeys|
+             ];
```

Comment:
Question: Is this selection by index of History of Digest calls?
I'm not familiar with `..` in Dafny,
but I presume it is slicing.

I am confused because `|old(cryptoPrimitives.History.Digest)|` is used twice...
OH, select from `oldIndex` to `oldIndex` + EDK count. 
Got it.

Reply:
Yes, it is slice.
Since there are some number of EDKs, I need to get all of them.
http://dafny.org/dafny/DafnyRef/DafnyRef#sec-other-sequence-expressions